### ISSUE TITLE
Optimize simple IFs as ternary with TUples

### ISSUE THREAD

**Issue by Daynvheur** [dotnet/roslyn#82644]
**Brief description:**

I would appreciate to transform more `if`s into ternaries when the condition are simple enough.

**Languages applicable:**

I think it is mainly C#, but VB may benefit from this aswell?

**Code example that the analyzer should report:**

Original code snippet to replace:
```cs
        private  string GetStringValueFromFile(string section, string variable, string defaultValue)
        {
            string toReturn;
            string defaultvalue = "Read value:";
            string value = GetValue(section, variable); // Reads from external file
            if (!string.IsNullOrWhiteSpace(value))
                toReturn = value;
            else
            {
                defaultvalue = "Set by default value :";
                toReturn = defaultValue;
            }
            Log.Write("LogHelper",  $"{nameof(GetStringValueFromFile)}: {variable}, {defaultValue}, {value}");

            return toReturn;
        }
```

**Additional information:**

Optimization in 3 steps (manually done here).
Unpacking the branches (defaultValue gets its allocation moved in the relevant branch):
```cs
        private  string GetStringValueFromFile(string section, string variable, string defaultValue)
        {
            string toReturn;
            string defaultvalue;
            string value = GetValue(section, variable); // Reads from external file
            if (!string.IsNullOrWhiteSpace(value))
            {
                defaultvalue = "Read value:";
                toReturn = value;
            }
            else
            {
                defaultvalue = "Set by default value :";
                toReturn = defaultValue;
            }
            Log.Write("LogHelper",  $"{nameof(GetStringValueFromFile)}: {variable}, {defaultValue}, {value}");

            return toReturn;
        }
```

TUplify the branches (allocation of both strings in one line):
```cs
        private  string GetStringValueFromFile(string section, string variable, string defaultValue)
        {
            string toReturn;
            string defaultvalue;
            string value = GetValue(section, variable); // Reads from external file
            if (!string.IsNullOrWhiteSpace(value))
            {
                (defaultvalue, toReturn) = ("Read value:", value);
            }
            else
            {
                (defaultvalue, toReturn) = ("Set by default value :", defaultValue);
            }
            Log.Write("LogHelper",  $"{nameof(GetStringValueFromFile)}: {variable}, {defaultValue}, {value}");

            return toReturn;
        }
```

Ternarify the branches (same as current ternaries, but applied to an extended case):
```cs
        private  string GetStringValueFromFile(string section, string variable, string defaultValue)
        {
            string value = GetValue(section, variable); // Reads from external file
            (string defaultvalue, string toReturn) = (!string.IsNullOrWhiteSpace(value)) ? ("Read value:", value)  : ("Set by default value :", defaultValue);
            Log.Write("LogHelper",  $"{nameof(GetStringValueFromFile)}: {variable}, {defaultValue}, {value}");

            return toReturn;
        }
```

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by Daynvheur** [comment]
For the note, IntelliSense's "Replace ternary into instructions" leads to invalid code:
```cs
        private  string GetStringValueFromFile(string section, string variable, string defaultValue)
        {
            string value = GetValue(section, variable); // Reads from external file
            if ((!string.IsNullOrWhiteSpace(value)))
            {
                (string defaultvalue, string toReturn) = ("Read value:", value);
            }
            else
            {
                (string defaultvalue, string toReturn) = ("Set by default value :", defaultValue);
            }
            Log.Write("LogHelper",  $"{nameof(GetStringValueFromFile)}: {variable}, {defaultValue}, {value}"); // CS0103 for defaultValue

            return toReturn; // CS0103 for toReturn
        }
```

Instead, there needs to be an extraction of the left handside assignees into the previous step ("TUplify the branches") as presented above:
```cs
        private  string GetStringValueFromFile(string section, string variable, string defaultValue)
        {
            string defaultvalue, toReturn; // Here, the left-handside assignees are declared
            string value = GetValue(section, variable); // Reads from external file
            if (!string.IsNullOrWhiteSpace(value))
            {
                (defaultvalue, toReturn) = ("Read value:", value); // assignees use previously-declared variables
            }
            else
            {
                (defaultvalue, toReturn) = ("Set by default value :", defaultValue); // assignees use previously-declared variables
            }
            Log.Write("LogHelper",  $"{nameof(GetStringValueFromFile)}: {variable}, {defaultValue}, {value}");

            return toReturn;
        }
```

**Comment by CyrusNajmabadi** [comment]
You can def write such a refactoring using the Roslyn API.

**Comment by Daynvheur** [comment]
I don't know how to write such API.
And I would love that baked-in!

But I'm glad to know this is something doable.

**Comment by CyrusNajmabadi** [comment]
> I don't know how to write such API.

I would recommend using an agent like Claude and writing against the roslyn refactoring API.

**Comment by Daynvheur** [comment]
(Here is a notebook prompt for later, written in French; feel free review with translation if you think some part need refinement)
> Pour répondre au besoin exprimé sur la page `https://github.com/dotnet/roslyn/issues/82644`, je voudrais un flux de travail pour l'API Roslyn en plusieurs parties :
> - Détecter un if/else ayant des branches avec valeurs par défaut implicites hors-branches (vs avec valeurs explicites in-branche), par exemple `string a="";int b;if(true){a="mardi";b=0;}else{b=1;}`
> - Déplacer les affectations implicites dans les branches explicitement (attention, les branches doivent être explicites), et retirer les affectations implicites hors-branches, par exemple `string a;int b;if(true){a="mardi";b=0;}else{b=1;a="";}`
> - Détecter les branches pouvant être "tuplifiées" en terme d'affectations, si possible en réordonnant les affectations pour obtenir des ordonnancements cohérents entre branches, par exemple `string a;int b;if(true){(a,b)=("mardi",0);}else{(a,b)=("", 1);}`
> - Appliquer la ternarification sur les branches ayant un tuple reconnu, par exemple `(string a;int b)=true?("mardi",0):("",1);`
> Précise si possible les contraintes et garde-fous consolidés, ainsi qu'une check-liste "prête à tester"

**Comment by Daynvheur** [comment]
Roslyn's existing API relevant to the situation:
- `UseConditionalExpressionForAssignmentHelpers.TryMatchPattern`: transforms regular `if(flag){x="";}else{x=".";}` into `x=flag?"":"."`
- `UseConditionalExpressionForReturnHelpers.TryMatchPattern`: transforms regular `if(flag){return "";}else{return ".";}` into `return "":".";`
- `UseExplicitTupleNameDiagnosticAnalyzer.AnalyzeOperation`: evaluates TUples via `field.ContainingType.IsTupleType`
